File: ios/Plugin/SubscriptionManagerPlugin.m
#import <Foundation/Foundation.h>
#import <Capacitor/Capacitor.h>

// Define the plugin using the CAP_PLUGIN Macro, and
// each method the plugin supports using the CAP_PLUGIN_METHOD macro.
CAP_PLUGIN(SubscriptionManagerPlugin, "SubscriptionManager",
           CAP_PLUGIN_METHOD(subscribe, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(hasValidSubscription, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(getSubscription, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(showManageSubscriptions, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(restoreSubscription, CAPPluginReturnPromise);
)

File: ios/Plugin/SubscriptionManagerPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(SubscriptionManagerPlugin)
public class SubscriptionManagerPlugin: CAPPlugin {
    private let implementation = SubscriptionManager()
    private let transactionObserver = TransactionObserver()
    
    @objc func subscribe(_ call: CAPPluginCall) {
        Task {
            let productId = call.getString("productId") ?? ""
            await implementation.subscribe(productId)
            call.resolve()
        }
    }
    
    @objc func hasSubscription(_ call: CAPPluginCall) {
        let productId = call.getString("productId") ?? ""
        Task {
            let ret = await implementation.hasSubscription(productId)
            call.resolve([
                "hasSubscription": ret
            ])
        }
    }

    @objc func getSubscription(_ call: CAPPluginCall) {
        let productId = call.getString("productId") ?? ""
        Task {
            let ret = await implementation.getSubscription(productId)
            call.resolve([
                "subscription": ret as Any
            ])
        }
    }

    @objc func showManageSubscriptions(_ call: CAPPluginCall) {
        Task {
            await implementation.showManageSubscriptions()
            call.resolve()
        }
    }
    
    @objc func restoreSubscription(_ call: CAPPluginCall) {
        Task {
            let productId = call.getString("productId") ?? ""
            await implementation.restoreSubscription(productId)
            call.resolve()
        }
    }
}
